library/spdm_requester_lib/libspdm_req_vendor_request.c

```diff
@@ -0,0 +1,258 @@
+
+    /* Copy Vendor id */
+    uint8_t* vendor_request = ((uint8_t *)spdm_request) + sizeof(spdm_vendor_defined_request_msg_t);
+    if (req_vendor_id_len != 0) {
+        libspdm_copy_mem(vendor_request, req_vendor_id_len, req_vendor_id, req_vendor_id_len);
+        vendor_request += req_vendor_id_len;
+    }
+
+    /* Copy request_len */
+    libspdm_copy_mem(vendor_request, sizeof(uint16_t), &req_size, sizeof(uint16_t));
+    vendor_request += sizeof(uint16_t);
+
+    /* Copy payload */
+    if (req_size != 0) {
+        libspdm_copy_mem(vendor_request, req_size, req_data, req_size);
+    }
+
+    spdm_request_size = sizeof(spdm_vendor_defined_request_msg_t) +
+                        req_vendor_id_len + sizeof(uint16_t) + req_size;
+
+    /* -=[Send Request Phase]=- */
+    status =
+        libspdm_send_spdm_request(spdm_context, session_id, spdm_request_size, spdm_request);
+    if (LIBSPDM_STATUS_IS_ERROR(status)) {
+        libspdm_release_sender_buffer (spdm_context);
+        status = LIBSPDM_STATUS_SEND_FAIL;
+        goto done;
+    }
+    libspdm_release_sender_buffer (spdm_context);
+    spdm_request = (void *)spdm_context->last_spdm_request;
+
+    /* -=[Receive Response Phase]=- */
+    status = libspdm_acquire_receiver_buffer (spdm_context, &message_size, (void **)&message);
+    if (LIBSPDM_STATUS_IS_ERROR(status)) {
+        return status;
+    }
+    LIBSPDM_ASSERT (message_size >= transport_header_size);
+    spdm_response = (void *)(message);
+    spdm_response_size = message_size;
+
+    libspdm_zero_mem(spdm_response, spdm_response_size);
+    status = libspdm_receive_spdm_response(spdm_context, session_id,
+                                           &spdm_response_size,
+                                           (void **)&spdm_response);
+
+    if (LIBSPDM_STATUS_IS_ERROR(status)) {
+        status = LIBSPDM_STATUS_RECEIVE_FAIL;
+        goto done;
+    }
+
+    /* -=[Validate Response Phase]=- */
+    /* check response buffer size at least spdm response default header plus
+     * number of bytes required by vendor id and 2 bytes for response payload size */
+    if (spdm_response_size < sizeof(spdm_vendor_defined_response_msg_t) +
+        spdm_response->vendor_id_len + sizeof(uint16_t)) {
+        status = LIBSPDM_STATUS_INVALID_MSG_SIZE;
+        goto done;
+    }
+    if (spdm_response->header.spdm_version != spdm_request->header.spdm_version) {
+        status = LIBSPDM_STATUS_INVALID_MSG_FIELD;
+        goto done;
+    }
+    if (spdm_response->header.request_response_code == SPDM_ERROR) {
+        status = libspdm_handle_error_response_main(
+            spdm_context, session_id,
+            &spdm_response_size,
+            (void **)&spdm_response, SPDM_VENDOR_DEFINED_REQUEST,
+            SPDM_VENDOR_DEFINED_RESPONSE);
+        if (LIBSPDM_STATUS_IS_ERROR(status)) {
+            goto done;
+        }
+    } else if (spdm_response->header.request_response_code != SPDM_VENDOR_DEFINED_RESPONSE) {
```

Comment:
At least as far as the specification is concerned, I don't think this is strictly illegal. I filed https://github.com/DMTF/SPDM-WG/issues/3309 for clarification. But the header documentation should state that a `VENDOR_DEFINED_RESPONSE` is expected by the Requester.